pkg/clients/compute.go

```diff
@@ -72,7 +86,18 @@ func NewComputeClient(providerClient *gophercloud.ProviderClient, providerClient
 	if err != nil {
 		return nil, fmt.Errorf("failed to create compute service client: %v", err)
 	}
-	compute.Microversion = NovaMinimumMicroversion
+
+	// Make an initial version negotiation.
+	// The caller can redo this later with more restrictions via RequireMicroversion.
+	supportedVersions := []*utils.Version{}
+	for k := range NovaSupportedVersions {
+		supportedVersions = append(supportedVersions, NovaSupportedVersions[k])
+	}
+	version, _, err := utils.ChooseVersion(providerClient, supportedVersions)
+	if err != nil {
+		return nil, fmt.Errorf("failed to negotiation compute service version: %w", err)
+	}
+	compute.Microversion = version.ID
```

Comment:
Hmm I'm not sure I understand. If we have deterministic microversion and no negotiation, then how would this return failure? Then we end up with the same as we already have where CAPO says the microversion is X and then we will see when making the first API call if that works or not. With negotiation we get to know this already here.
I can accept negotiating with a single microversion (so it is that microversion or nothing), but then I still think it is better to do the negotiation than just setting it without checking if it is supported.

Reply:
2 things:

I anticipate that we'd fetch the maximum supported microversion from Nova when initially creating the service client and store it in the service client. We would then use this every time we needed to check that the required microversion was supported. i.e.

On instantiating the service client we check that microversion 38 is supported because that's the minimum. This is the microversion that all code without a higher requirement will run against.

When adding tags to a server before creation we'd do something like:
```go
  createClient, err := compute.WithMicroversion(TAGS)
  if err != nil {
    createClient = compute
  } else {
    ... add tags to server create opts ...
  }
```